apypie/foreman.py

```diff
@@ -0,0 +1,227 @@
+        if not ignore_errors and task['result'] != 'success':
+            msg = f"Task {task['action']}({task['id']}) did not succeed. Task information: {task['humanized']['errors']}"
+            raise ForemanApiException(msg=msg)
+        return task
+
+    def show(self, resource: str, resource_id: int, params: Optional[dict] = None) -> Optional[dict]:
+        """
+        Execute the ``show`` action on an entity.
+
+        :param resource: Plural name of the api resource to show
+        :param resource_id: The ID of the entity to show
+        :param params: Lookup parameters (i.e. parent_id for nested entities)
+
+        :return: The entity
+        """
+        payload = {'id': resource_id}
+        if params:
+            payload.update(params)
+        return self.resource_action(resource, 'show', payload)
+
+    def list(self, resource: str, search: Optional[str] = None, params: Optional[dict] = None) -> list:
+        """
+        Execute the ``index`` action on an resource.
+
+        :param resource: Plural name of the api resource to show
+        :param search: Search string as accepted by the API to limit the results
+        :param params: Lookup parameters (i.e. parent_id for nested entities)
+
+        :return: List of results
+        """
+        payload: dict = {'per_page': PER_PAGE}
+        if search is not None:
+            payload['search'] = search
+        if params:
+            payload.update(params)
+
+        result = self.resource_action(resource, 'index', payload)
+        if result:
+            return result['results']
+        return []
+
+    def create(self, resource: str, desired_entity: dict, params: Optional[dict] = None) -> Optional[dict]:
+        """
+        Create entity with given properties
+
+        :param resource: Plural name of the api resource to manipulate
+        :param desired_entity: Desired properties of the entity
+        :param params: Lookup parameters (i.e. parent_id for nested entities)
+
+        :return: The new current state of the entity
+        """
+        payload = desired_entity.copy()
+        if params:
+            payload.update(params)
+        return self.resource_action(resource, 'create', payload)
+
+    def update(self, resource: str, desired_entity: dict, params: Optional[dict] = None) -> Optional[dict]:
+        """
+        Update entity with given properties
+
+        :param resource: Plural name of the api resource to manipulate
+        :param desired_entity: Desired properties of the entity
+        :param params: Lookup parameters (i.e. parent_id for nested entities)
+
+        :return: The new current state of the entity
+        """
+        payload = desired_entity.copy()
+        if params:
+            payload.update(params)
+        return self.resource_action(resource, 'update', payload)
+
+    def delete(self, resource: str, current_entity: dict, params: Optional[dict] = None) -> None:  # pylint: disable=useless-return
```

Comment:
should this be called `delete` (as that's the HTTP verb that gets executed) or `destroy` as that's the action Foreman defines?

Reply:
`delete` feels more Pythonic to me